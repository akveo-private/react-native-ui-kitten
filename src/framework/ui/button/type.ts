export interface ButtonAlignment {
  rawValue: string;

  flex(): FlexAlignment;
}

export type FlexAlignment = 'row' | 'row-reverse';

export class ButtonAlignments {
  static LEFT: ButtonAlignment = new class implements ButtonAlignment {
    rawValue: string = 'left';

    flex(): FlexAlignment {
      return 'row';
    }
  };

  static RIGHT: ButtonAlignment = new class implements ButtonAlignment {
    rawValue: string = 'right';

    flex(): FlexAlignment {
      return 'row-reverse';
    }
  };

  static parse(value: string | ButtonAlignment, fallback?: ButtonAlignment): ButtonAlignment | undefined {
    if (ButtonAlignments.typeOf(value)) {
      return value;
    }

    return ButtonAlignments.parseString(value, fallback);
  }

  private static parseString(rawValue: string, fallback?: ButtonAlignment): ButtonAlignment | undefined {
    switch (rawValue) {
      case ButtonAlignments.LEFT.rawValue:
        return ButtonAlignments.LEFT;
      case ButtonAlignments.RIGHT.rawValue:
        return ButtonAlignments.RIGHT;
      default:
        return fallback;
    }
  }

  private static typeOf(value: any): value is ButtonAlignment {
    const { rawValue } = (<ButtonAlignment>value);

    return rawValue !== undefined;
  }
}
