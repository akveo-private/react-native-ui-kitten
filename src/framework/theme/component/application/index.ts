export {
  ApplicationProvider,
  Props as ApplicationProviderProps,
} from './applicationProvider.component';
