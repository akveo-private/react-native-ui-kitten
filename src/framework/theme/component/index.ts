export { ThemeProvider } from './themeProvider.component';
export { withTheme } from './themeConsumer.component';
export { withThemedStyles } from './themeConsumerStyled.component';
export { ThemeType } from './type';
