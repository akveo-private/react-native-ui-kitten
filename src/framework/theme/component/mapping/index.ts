export * from './type';
export {
  MappingProvider as ThemeMappingProvider,
  Props as ThemeMappingProviderProps,
} from './mappingProvider.component';
