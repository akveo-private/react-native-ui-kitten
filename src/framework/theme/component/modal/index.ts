export * from './modal.component';
export * from './modalPanel.component';
