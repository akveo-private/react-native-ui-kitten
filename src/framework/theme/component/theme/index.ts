export * from './type';
export {
  ThemeProvider,
  Props as ThemeProviderProps,
} from './themeProvider.component';
export {
  withStyles,
  Props as ThemedComponentProps,
} from './themeConsumer.component';
