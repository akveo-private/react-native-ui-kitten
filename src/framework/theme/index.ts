export {
  StyleProvider,
  styled,
  ThemeProvider,
  withStyles,

  StyleProviderProps,
  StyledComponentProps,
  ThemeProviderProps,
  ThemedComponentProps,

  ThemeType,
  ThemeMappingType,
  ComponentMappingType,
  TokenType,
  ThemedStyleType,
  StyleSheetType,
  StyleType,
} from './component';

export {
  APPEARANCE_DEFAULT,
} from './service';

