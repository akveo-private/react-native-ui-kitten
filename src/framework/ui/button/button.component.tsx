import React from 'react';
import {
  TouchableOpacity,
  TouchableOpacityProps,
  GestureResponderEvent,
  StyleSheet,
  ImageSourcePropType,
  Image,
  ImageProps,
} from 'react-native';
import {
  StyledComponentProps,
  StyleType,
  Interaction,
  styled,
} from '@kitten/theme';
import {
  Text as TextComponent,
  Props as TextProps,
} from '../text/text.component';
import {
  ButtonAlignment,
  ButtonAlignments,
} from './type';

interface ButtonProps {
  icon?: ImageSourcePropType;
  status?: string;
  size?: string;
  alignment?: string | ButtonAlignment;
  children?: React.ReactText;
}

const Text = styled<TextComponent, TextProps>(TextComponent);

export type Props = ButtonProps & StyledComponentProps & TouchableOpacityProps;

const ALIGNMENT_DEFAULT: ButtonAlignment = ButtonAlignments.LEFT;

export class Button extends React.Component<Props> {

  static defaultProps: Partial<Props> = {
    status: 'primary',
    size: 'medium',
  };

  private onPress = (event: GestureResponderEvent) => {
    if (this.props.onPress) {
      this.props.onPress(event);
    }
  };

  private onPressIn = (event: GestureResponderEvent) => {
    this.props.dispatch([Interaction.ACTIVE]);

    if (this.props.onPressIn) {
      this.props.onPressIn(event);
    }
  };

  private onPressOut = (event: GestureResponderEvent) => {
    this.props.dispatch([]);

    if (this.props.onPressOut) {
      this.props.onPressOut(event);
    }
  };

  private getComponentStyle = (style: StyleType): StyleType => {
    const { text, icon, ...container } = style;
    const alignment: ButtonAlignment = ButtonAlignments.parse(this.props.alignment, ALIGNMENT_DEFAULT);

    return {
      container: {
        ...container,
        flexDirection: alignment.flex(),
      },
      text: text,
      icon: icon,
    };
  };

  private createTextElement = (style: StyleType): React.ReactElement<TextProps> => {
    const { children: text } = this.props;

    return (
      <Text
        style={[style, strictStyles.text]}
        key={1}>
        {text}
      </Text>
    );
  };

  private createImageElement = (style: StyleType): React.ReactElement<ImageProps> => {
    const { icon: image } = this.props;

    return (
      <Image
        key={0}
        style={[style, strictStyles.icon]}
        source={image}
      />
    );
  };

  private createComponentChildren = (style: StyleType): React.ReactNode => {
    const { icon, children } = this.props;

    const hasIcon: boolean = icon !== undefined;
    const hasText: boolean = children !== undefined;

    return [
      hasIcon ? this.createImageElement(style.icon) : undefined,
      hasText ? this.createTextElement(style.text) : undefined,
    ];
  };

  public render(): React.ReactElement<TouchableOpacityProps> {
    const { style, themedStyle, ...derivedProps } = this.props;
    const { container, ...componentStyles } = this.getComponentStyle(themedStyle);
    const componentChildren: React.ReactNode = this.createComponentChildren(componentStyles);

    return (
      <TouchableOpacity
        {...derivedProps}
        style={[style, container, strictStyles.container]}
        activeOpacity={1.0}
        onPress={this.onPress}
        onPressIn={this.onPressIn}
        onPressOut={this.onPressOut}>
        {componentChildren}
      </TouchableOpacity>
    );
  }
}

const strictStyles = StyleSheet.create({
  container: {
    justifyContent: 'space-evenly',
    alignItems: 'center',
  },
  text: {},
  icon: {},
});
