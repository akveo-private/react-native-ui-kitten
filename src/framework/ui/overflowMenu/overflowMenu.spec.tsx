import React from 'react';
import {
  fireEvent,
  render,
  RenderAPI,
} from 'react-native-testing-library';
import { ApplicationProvider } from '@kitten/theme';
import {
  OverflowMenuItemType,
  OverflowMenu,
  OverflowMenuProps,
} from './overflowMenu.component';
import { Button } from '../button/button.component';
import {
  mapping,
  theme,
} from '../support/tests';

interface State {
  menuVisible: boolean;
  selectedIndex: number;
}

interface ComponentProps {
  onSelectChecker?: () => void;
}

type Props = ComponentProps & Partial<OverflowMenuProps>;

class TestApplication extends React.Component<Props, State> {

  public state: State = {
    menuVisible: false,
    selectedIndex: null,
  };

  private setMenuVisible = (): void => {
    const menuVisible: boolean = !this.state.menuVisible;

    this.setState({ menuVisible });
  };

  private onSelect = (selectedIndex: number): void => {
    this.props.onSelectChecker && this.props.onSelectChecker();
    this.setState({ selectedIndex }, this.setMenuVisible);
  };

  public render(): React.ReactNode {
    const { data } = this.props;
    const { menuVisible, selectedIndex } = this.state;

    return (
      <ApplicationProvider
        mapping={mapping}
        theme={theme}>
        <OverflowMenu
          data={data}
          visible={menuVisible}
          selectedIndex={selectedIndex}
          onBackdropPress={this.setMenuVisible}
          onSelect={this.onSelect}>
          <Button onPress={this.setMenuVisible}>Show</Button>
        </OverflowMenu>
      </ApplicationProvider>
    );
  }
}

describe('@overflow-menu: component checks', () => {

  const defaultItems: OverflowMenuItemType[] = [
    { title: 'Option 1' },
    { title: 'Option 2' },
    { title: 'Option 3' },
  ];

  it('* menu-item visible prop check', () => {
    const application: RenderAPI = render(
      <TestApplication
        data={defaultItems}
      />,
    );

    fireEvent.press(application.getByType(Button));

    const { visible } = application.getByType(OverflowMenu).props;
    expect(visible).toBe(true);
  });

  it('* menu-item onSelect prop check', () => {
    const onSelectChecker = jest.fn();
    const application: RenderAPI = render(
      <TestApplication
        data={defaultItems}
        onSelectChecker={onSelectChecker}
      />,
    );

    fireEvent.press(application.getByType(Button));
    fireEvent.press(application.getAllByText('Option 1')[0]);

    expect(onSelectChecker).toHaveBeenCalled();
  });

  it('* menu-item onSelect/selectedIndex prop check', () => {
    const application: RenderAPI = render(
      <TestApplication
        data={defaultItems}
      />,
    );

    fireEvent.press(application.getByType(Button));
    fireEvent.press(application.getAllByText('Option 2')[0]);

    const { selectedIndex } = application.getByType(OverflowMenu).props;

    expect(selectedIndex).toBe(1);
  });

});
