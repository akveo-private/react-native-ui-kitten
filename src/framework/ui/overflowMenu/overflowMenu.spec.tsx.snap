// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@overflow-menu: component checks * component renders properly 1`] = `
<Wrapper
  forwardedRef={null}
  items={
    Array [
      Object {
        "icon": [Function],
        "text": "Menu Item 1",
        "textStyle": Object {
          "color": "blue",
          "fontSize": 24,
        },
      },
      Object {
        "disabled": true,
        "icon": [Function],
        "text": "Menu Item 2",
      },
      Object {
        "text": "Menu Item 3",
      },
      Object {
        "icon": [Function],
        "text": "Menu Item 4",
      },
    ]
  }
  onRequestClose={[Function]}
  visible={true}
>
  <View />
</Wrapper>
`;

exports[`@overflow-menu: component checks * single menu-item 1`] = `
<Wrapper
  forwardedRef={null}
  items={
    Array [
      Object {
        "icon": [Function],
        "text": "Menu Item 1",
        "textStyle": Object {
          "color": "blue",
          "fontSize": 24,
        },
      },
    ]
  }
  onRequestClose={[Function]}
  visible={true}
>
  <View />
</Wrapper>
`;

exports[`@overflow-menu-item: component checks * menu item active checks 1`] = `
<Wrapper
  forwardedRef={null}
  onLongPress={[MockFunction]}
  onPressIn={
    [MockFunction] {
      "calls": Array [
        Array [
          undefined,
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onPressOut={
    [MockFunction] {
      "calls": Array [
        Array [
          undefined,
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  text="Test Menu Item"
/>
`;

exports[`@overflow-menu-item: component checks * menu item active checks 2`] = `
<Wrapper
  forwardedRef={null}
  onLongPress={[MockFunction]}
  onPressIn={
    [MockFunction] {
      "calls": Array [
        Array [
          undefined,
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onPressOut={
    [MockFunction] {
      "calls": Array [
        Array [
          undefined,
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  text="Test Menu Item"
/>
`;

exports[`@overflow-menu-item: component checks * menu item with "set-1" props 1`] = `
<Wrapper
  disabled={true}
  forwardedRef={null}
  icon={[Function]}
  onPress={[Function]}
  text="Test Menu Item"
/>
`;

exports[`@overflow-menu-item: component checks * menu item with "set-2" props 1`] = `
<Wrapper
  disabled={false}
  forwardedRef={null}
  onPress={[Function]}
  text="Test Menu Item"
/>
`;
